File: q4/C++/q4.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* reConstructBinaryTree(vector<int> pre,vector<int> vin) {
        TreeNode* root = nullptr;
        if (pre.empty() || vin.empty() || pre.size() != vin.size())
            return root;
        
        root = constructT(pre, vin, 0, pre.size() - 1, 0, vin.size() - 1);
        
        return root;
    }
    
    TreeNode* constructT(const vector<int>& pre,
                         const vector<int>& vin,
                        int preStart, int preEnd,
                         int vinStart, int vinEnd) {
        if (preStart > preEnd || vinStart > vinEnd)
            return nullptr;
        
        TreeNode* root = new TreeNode(pre[preStart]);
            
        for (int i = vinStart; i <= vinEnd; i++) {
            if (pre[preStart] == vin[i]){
                root->left = constructT(pre, vin,
                                        preStart + 1, preStart + i - vinStart,
                                        vinStart, i - 1);
                root->right = constructT(pre, vin,
                                         preStart + i - vinStart + 1, preEnd,
                                         i + 1, vinEnd);
            }
        }
        
        return root;
    }
};

File: q4/JAVA/q4.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode constructTree(int [] pre, int [] in, 
                                  int preStart, int preEnd,
                                 int inStart, int inEnd) {
        if (preStart > preEnd || inStart > inEnd) {
            return null;
        }
        TreeNode root = new TreeNode(pre[preStart]);
        for (int i = inStart; i <= inEnd; i++) {
            if (pre[preStart] == in[i]) {
                root.left = constructTree(pre, in,
                                           preStart + 1, preStart + i - inStart,
                                          inStart, i - 1);
                root.right = constructTree(pre, in,
                                           preStart + i - inStart + 1, preEnd,
                                           i + 1, inEnd);
                return root;
            }
        }
        return root;
    }
    public TreeNode reConstructBinaryTree(int [] pre,int [] in) {
        if (pre.length == 0 || in.length == 0 || pre.length != in.length) {
            return null;
        }
        
        return constructTree(pre, in, 0, pre.length - 1, 0, in.length - 1);
    }
}